sui-execution/latest/sui-adapter/src/package_unification.rs

```diff
@@ -0,0 +1,379 @@
+    pub fn try_compute_unified_linkage(
+        &self,
+        linking_config: LinkageConfig,
+    ) -> anyhow::Result<BTreeSet<ObjectID>> {
+        let mut unification_table = BTreeMap::new();
+
+        // Any packages that are being upgraded cannot be called in the transaction
+        for object_id in self.upgrading_packages.iter() {
+            Self::add_and_unify(
+                &mut unification_table,
+                &self.all_packages,
+                object_id,
+                |pkg| ConflictResolution::Never(pkg.id()),
+            )?;
+        }
+
+        // Linkages for packages that are to be published must be exact.
+        for object_id in self.publication_linkages.iter() {
+            Self::add_and_unify(
+                &mut unification_table,
+                &self.all_packages,
+                object_id,
+                |pkg| ConflictResolution::Exact(pkg.version(), pkg.id()),
+            )?;
+        }
+
+        for object_id in self.entry_functions.iter() {
+            Self::add_and_unify(
+                &mut unification_table,
+                &self.all_packages,
+                object_id,
+                |pkg| ConflictResolution::Exact(pkg.version(), pkg.id()),
+            )?;
+
+            // transitive closure of entry functions are fixed
+            for dep_id in self.all_packages[object_id]
+                .linkage_table()
+                .values()
+                .map(|info| &info.upgraded_id)
+            {
+                Self::add_and_unify(&mut unification_table, &self.all_packages, dep_id, |pkg| {
+                    ConflictResolution::Exact(pkg.version(), pkg.id())
+                })?;
+            }
+        }
+
+        // Types can be fixed or not depending on config.
+        for object_id in self.types.iter() {
+            Self::add_and_unify(
+                &mut unification_table,
+                &self.all_packages,
+                object_id,
+                |pkg| {
+                    if linking_config.fix_types {
+                        ConflictResolution::Exact(pkg.version(), pkg.id())
+                    } else {
+                        ConflictResolution::AtLeast(pkg.version(), *object_id)
+                    }
+                },
+            )?;
+        }
+
+        // Top level functions can be fixed or not depending on config. But they won't ever
+        // transitively fix their dependencies.
+        for object_id in self.non_entry_functions.iter() {
+            Self::add_and_unify(
+                &mut unification_table,
+                &self.all_packages,
+                object_id,
+                |pkg| {
+                    if linking_config.fix_top_level_functions {
+                        ConflictResolution::Exact(pkg.version(), pkg.id())
+                    } else {
+                        ConflictResolution::AtLeast(pkg.version(), *object_id)
+                    }
+                },
+            )?;
+        }
```

Comment:
A small nit for logic like this: consider moving it onto the `linking_config`, as:

```
        for object_id in self.non_entry_functions.iter() {
            Self::add_and_unify(
                &mut unification_table,
                &self.all_packages,
                object_id,
                |pkg| linking_config.generate_top_level_function_constraint(pkg.version(), pkg.id(), object_id)
            )?;
        }
```
That way as the flags grow, we do not need to update their usage across the file; we can just ask the config what it should be.